src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java

```diff
@@ -0,0 +1,173 @@
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.SSTableMultiWriter;
+import org.apache.cassandra.io.sstable.format.big.BigTableBlockWriter;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.streaming.ProgressInfo;
+import org.apache.cassandra.streaming.StreamReceiver;
+import org.apache.cassandra.streaming.StreamSession;
+import org.apache.cassandra.streaming.messages.StreamMessageHeader;
+
+import static java.lang.String.format;
+import static org.apache.cassandra.utils.FBUtilities.prettyPrintMemory;
+
+/**
+ * CassandraBlockStreamReader reads SSTable off the wire and writes it to disk.
+ */
+public class CassandraBlockStreamReader implements IStreamReader
+{
+    private static final Logger logger = LoggerFactory.getLogger(CassandraBlockStreamReader.class);
+
+    private final TableId tableId;
+    private final StreamSession session;
+    private final CassandraStreamHeader header;
+    private final int fileSequenceNumber;
+
+    public CassandraBlockStreamReader(StreamMessageHeader messageHeader, CassandraStreamHeader streamHeader, StreamSession session)
+    {
+        if (session.getPendingRepair() != null)
+        {
+            // we should only ever be streaming pending repair sstables if the session has a pending repair id
+            if (!session.getPendingRepair().equals(messageHeader.pendingRepair))
+                throw new IllegalStateException(format("Stream Session & SSTable (%s) pendingRepair UUID mismatch.", messageHeader.tableId));
+        }
+
+        this.header = streamHeader;
+        this.session = session;
+        this.tableId = messageHeader.tableId;
+        this.fileSequenceNumber = messageHeader.sequenceNumber;
+    }
+
+    /**
+     * @param inputPlus where this reads data from
+     * @return SSTable transferred
+     * @throws IOException if reading the remote sstable fails. Will throw an RTE if local write fails.
+     */
+    @SuppressWarnings("resource") // input needs to remain open, streams on top of it can't be closed
+    @Override
+    public SSTableMultiWriter read(DataInputPlus inputPlus) throws IOException
+    {
+        ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(tableId);
+        if (cfs == null)
+        {
+            // schema was dropped during streaming
+            throw new IOException("Table " + tableId + " was dropped during streaming");
+        }
+
+        ComponentManifest manifest = header.componentManifest;
+        long totalSize = manifest.totalSize();
+
+        logger.debug("[Stream #{}] Started receiving sstable #{} from {}, size = {}, table = {}",
+                     session.planId(),
+                     fileSequenceNumber,
+                     session.peer,
+                     prettyPrintMemory(totalSize),
+                     cfs.metadata());
+
+        BigTableBlockWriter writer = null;
```

Comment:
It's likely that we'll have more formats in near future. I don't suggest we make this generic, until then, but for now at least validate that the format in the header is `BIG` perhaps?

Reply:
Done